capsules/src/sht3x.rs

```diff
@@ -0,0 +1,245 @@
+                buffer[1] = ((Registers::MEASHIGHREP as u16) & 0xff) as u8;
+
+                self.i2c.write(buffer, 2);
+
+                ReturnCode::SUCCESS
+            },
+        )
+    }
+}
+
+impl<'a, A: Alarm<'a>> time::AlarmClient for SHT3x<'a, A> {
+    fn alarm(&self) {
+        let state = self.state.get();
+        match state {
+            State::Read => {
+                self.state.set(State::ReadData);
+                self.buffer.take().map_or_else(
+                    || panic!("SHT3x No buffer available!"),
+                    |buffer| {
+                        let _res = self.i2c.read(buffer, 6);
+                    },
+                );
+            }
+            _ => {
+                // This should never happen
+                panic!("SHT31 Invalid alarm!");
+            }
+        }
+    }
+}
+
+impl<'a, A: Alarm<'a>> i2c::I2CClient for SHT3x<'a, A> {
+    fn command_complete(&self, buffer: &'static mut [u8], error: i2c::Error) {
+        match error {
+            i2c::Error::CommandComplete => {
+                let state = self.state.get();
+
+                match state {
+                    State::ReadData => {
+                        if self.read_temp.get() == true {
+                            self.read_temp.set(false);
+                            if crc8(&buffer[0..2]) == buffer[2] {
+                                let mut stemp = buffer[0] as u32;
+                                stemp = stemp << 8;
+                                stemp = stemp | buffer[1] as u32;
+                                stemp = ((4375 * stemp) >> 14) - 4500;
+                                self.temperature_client
+                                    .map(|cb| cb.callback(stemp as usize));
+                            } else {
+                                self.temperature_client.map(|cb| cb.callback(usize::MAX));
+                            }
+                        }
+                        if self.read_hum.get() == true {
+                            self.read_hum.set(false);
+                            if crc8(&buffer[3..5]) == buffer[5] {
+                                let mut shum = buffer[3] as u32;
+                                shum = shum << 8;
+                                shum = shum | buffer[4] as u32;
+                                shum = (625 * shum) >> 12;
+                                self.humidity_client.map(|cb| cb.callback(shum as usize));
+                            } else {
+                                self.humidity_client.map(|cb| cb.callback(usize::MAX));
+                            }
+                        }
+                        self.buffer.replace(buffer);
+                        self.state.set(State::Idle);
+                    }
+                    State::Read => {
+                        self.buffer.replace(buffer);
+                        let interval = A::ticks_from_ms(20);
+                        self.alarm.set_alarm(self.alarm.now(), interval);
```

Comment:
can you explain why you set an alarm here, rather than beginning the next operation immediately?

Reply:
The sensor needs 20 ms between the measurement command and the read command.